frame/oracle/src/tests.rs

```diff
@@ -20,205 +20,462 @@ use sp_runtime::{
-		const BLOCK_INTERVAL: u64 = 5;
-		const REWARD: u64 = 5;
-		const SLASH: u64 = 5;
-
-		// passes
-		let account_2 = get_account_2();
-		assert_ok!(Oracle::add_asset_and_info(
-			Origin::signed(account_2),
-			ASSET_ID,
-			THRESHOLD,
-			MIN_ANSWERS,
-			MAX_ANSWERS,
-			BLOCK_INTERVAL,
-			REWARD,
-			SLASH
-		));
-
-		// does not increment if exists
-		assert_ok!(Oracle::add_asset_and_info(
-			Origin::signed(account_2),
-			ASSET_ID,
-			THRESHOLD,
-			MIN_ANSWERS,
-			MAX_ANSWERS,
-			BLOCK_INTERVAL,
-			REWARD,
-			SLASH
-		));
-		assert_eq!(Oracle::assets_count(), 1);
-
-		assert_ok!(Oracle::add_asset_and_info(
-			Origin::signed(account_2),
-			ASSET_ID + 1,
-			THRESHOLD,
-			MIN_ANSWERS,
-			MAX_ANSWERS,
-			BLOCK_INTERVAL,
-			REWARD,
-			SLASH
-		));
-
-		let asset_info = AssetInfo {
-			threshold: THRESHOLD,
-			min_answers: MIN_ANSWERS,
-			max_answers: MAX_ANSWERS,
-			block_interval: BLOCK_INTERVAL,
-			reward: REWARD,
-			slash: SLASH,
-		};
-		// id now activated and count incremented
-		assert_eq!(Oracle::asset_info(1), Some(asset_info));
-		assert_eq!(Oracle::assets_count(), 2);
-
-		// fails with non permission
-		let account_1 = get_account_1();
-		assert_noop!(
-			Oracle::add_asset_and_info(
-				Origin::signed(account_1),
-				ASSET_ID,
-				THRESHOLD,
-				MAX_ANSWERS,
-				MAX_ANSWERS,
-				BLOCK_INTERVAL,
-				REWARD,
-				SLASH
-			),
-			BadOrigin
-		);
-
-		assert_noop!(
-			Oracle::add_asset_and_info(
-				Origin::signed(account_2),
-				ASSET_ID,
-				THRESHOLD,
-				MAX_ANSWERS,
-				MIN_ANSWERS,
-				BLOCK_INTERVAL,
-				REWARD,
-				SLASH
-			),
-			Error::<Test>::MaxAnswersLessThanMinAnswers
-		);
-
-		assert_noop!(
-			Oracle::add_asset_and_info(
-				Origin::signed(account_2),
-				ASSET_ID,
-				Percent::from_percent(100),
-				MIN_ANSWERS,
-				MAX_ANSWERS,
-				BLOCK_INTERVAL,
-				REWARD,
-				SLASH
-			),
-			Error::<Test>::ExceedThreshold
-		);
-
-		assert_noop!(
-			Oracle::add_asset_and_info(
-				Origin::signed(account_2),
-				ASSET_ID,
-				THRESHOLD,
-				MIN_ANSWERS,
-				MAX_ANSWERS + 1,
-				BLOCK_INTERVAL,
-				REWARD,
-				SLASH
-			),
-			Error::<Test>::ExceedMaxAnswers
-		);
-
-		assert_noop!(
-			Oracle::add_asset_and_info(
-				Origin::signed(account_2),
-				ASSET_ID,
-				THRESHOLD,
-				0,
-				MAX_ANSWERS,
-				BLOCK_INTERVAL,
-				REWARD,
-				SLASH
-			),
-			Error::<Test>::InvalidMinAnswers
-		);
+use proptest::prelude::*;
+use composable_tests_helpers::{prop_assert_noop, prop_assert_ok};
+
+use sp_core::H256;
+
+prop_compose! {
+    fn asset_info()
+        (
+            min_answers in 1..MaxAnswerBound::get(),
+            max_answers in 1..MaxAnswerBound::get(),
+            block_interval in (StalePrice::get()+1)..(BlockNumber::MAX/16),
+            threshold in 0..100u8,
+            reward in 0..u64::MAX,
+            slash in 0..u64::MAX,
+        ) -> AssetInfo<Percent, BlockNumber, Balance> {
+            let min_answers = max_answers.saturating_sub(min_answers) + 1;
```

Comment:
i used prop_filter to filter out cases when max < min. As i see now you test exactly same case of 1 way to many times.

Reply:
@dzmitry-lahoda  Where did you use prop_filter like this? Could you link to it?

Reply:
https://github.com/search?q=org%3AComposableFi+prop_filter&type=code